Salesforce Connector 9.9 (#761)
mulesoft/docs-connectors: modules/ROOT/pages/salesforce/salesforce-connector-reference-99.adoc

| Consumer Key a| String |  Consumer key for Salesforce connected app |  | x
| Key Store a| String |  Path to key store used to sign data during authentication |  | x
| Store Password a| String |  Password of key store |  | x
| Principal a| String |  Username of desired Salesforce user to take action on behalf of. |  | x
| Token Endpoint a| String |  URL pointing to the server responsible for providing the authentication token. According to Salesforce it should be `\https://login.salesforce.com/services/oauth2/token`, or, if implementing for a community, `\https://acme.force.com/customers/services/oauth2/token` (where `acme.force.com/customers` is your community URL). |  `\https://login.salesforce.com/services/oauth2/token` |
| Audience Url a| String | The URL that identifies the authorization server as an intended audience. The authorization server must verify that it is an intended audience for the token.

> Use the authorization server's URL for the audience value: `/https://login.salesforce.com`, `\https://test.salesforce.com`, or `\https://community.force.com/customers` if implementing for a community. | https://login.salesforce.com |

Reviewer: The slash in front of https for login is / and the one for test is \
Not sure which one is right, but they should probably be the same.  ?

Reviewer: I think \ is the right way

| Token Endpoint a| String |  URL pointing to the server responsible for providing the authentication token. According to Salesforce it should be `\https://login.salesforce.com/services/oauth2/token`, or, if implementing for a community, `\https://acme.force.com/customers/services/oauth2/token` (where `acme.force.com/customers` is your community URL). |  `\https://login.salesforce.com/services/oauth2/token/` |
| Reconnection a| <<Reconnection>> |  When the application is deployed, a connectivity test is performed on all connectors. If set to true, deployment fails if the test doesn't pass after exhausting the associated reconnection strategy. |  |
|===
[[SfdcConfig_ConfigWithOauth]]
===== OAuth v2.0

====== Parameters

[%header,cols="20s,25a,30a,15a,10a"]
|===
| Name | Type | Description | Default Value | Required
| Host a| String |  Hostname of the proxy. If this property is not set then no proxy is used; otherwise, a proxy is used, but a proxy host must be specified. |  |
| Port a| Number |  Port of the proxy. If host is set then this property must be set and cannot be a negative number. |  |

> | Username a| String |  Username used to authenticate against the proxy. If this property is not set then no authentication is going to be used against the proxy, otherwise this value must be specified. |  |

Reviewer: ```suggestion
| Username a| String |  Username used to authenticate against the proxy. If this property is not set then no authentication is used against the proxy. |  |
```

| Can Clear Fields by Updating Field value to Null a| Boolean |  If false, then to clear a field, its name must be provided in an update request in the `fieldsToNull` field, otherwise, for clearing a field it can also be simply added to the request with the value `null`. Default value is `false`. |  false |
| Consumer Key a| String |  Consumer key for Salesforce connected app |  | x
| Key Store a| String |  Path to key store used to sign data during authentication |  | x
| Store Password a| String |  Password of key store |  | x
| Principal a| String |  Username of desired Salesforce user to take action on behalf of. |  | x
| Token Endpoint a| String |  URL pointing to the server responsible for providing the authentication token. According to Salesforce it should be `+https://login.salesforce.com/services/oauth2/token+`, or, if implementing for a community, `\https://acme.force.com/customers/services/oauth2/token` (where `acme.force.com/customers` is your community URL). |  `\https://login.salesforce.com/services/oauth2/token` |

> | Audience Url a| String |  The audience identifies the authorization server as an intended audience. The authorization server must verify that it is an intended audience for the token. <p> Use the authorization server's URL for the audience value: https://login.salesforce.com, https://test.salesforce.com, or https://community.force.com/customers if implementing for a community. | https://login.salesforce.com |

Reviewer: Do the URLs need to be escaped?

* SALESFORCE:RETRY_EXHAUSTED
* SALESFORCE:UNKNOWN

[[AbortJobV2]]
== Abort Job V2
`<salesforce:abort-job-v2>`

Aborts an ongoing Bulk API V2 Job. This call uses the Bulk API v2.

=== Parameters

[%header,cols="20s,25a,30a,15a,10a"]
|===
| Name | Type | Description | Default Value | Required
| Configuration | String | The name of the configuration to use. | | x
| Job id a| String |  The ID of the Job. |  | x
| Target Variable a| String |  The name of a variable that stores the operation's output |  |
| Target Value a| String |  An expression that evaluates the operation's output and stores the output in Target Variable |  #[payload] |
| Reconnection Strategy a| * <<reconnect>>
* <<reconnect-forever>> |  A retry strategy in case of connectivity errors |  |
|===

=== Output

[%autowidth.spread]
|===
| *Type* a| <<BulkJobState>>
|===

=== For Configurations

* <<SfdcConfig>>

=== Throws

* SALESFORCE:CONNECTIVITY
* SALESFORCE:INVALID_REQUEST_DATA
* SALESFORCE:INVALID_SESSION
* SALESFORCE:INVALID_STRUCTURE_FOR_INPUT_DATA
* SALESFORCE:RETRY_EXHAUSTED
* SALESFORCE:UNKNOWN

[[AbortQueryJob]]
== Abort Query Job
`<salesforce:abort-query-job>`

> Aborts a query job, based on its id.

Reviewer: ```suggestion
Aborts a query job, based on its ID.
```

=== Parameters

[%header,cols="20s,25a,30a,15a,10a"]
|===
| Name | Type | Description | Default Value | Required
| Configuration | String | The name of the configuration to use. | | x
| Query Job Id a| String |  The ID of the query job |  | x
| Target Variable a| String |  The name of a variable that stores the operation's output |  |
| Target Value a| String |  An expression that evaluates the operation's output and stores the output in Target Variable |  #[payload] |
| Reconnection Strategy a| * <<reconnect>>
* <<reconnect-forever>> |  A retry strategy in case of connectivity errors |  |
|===

=== Output

[%autowidth.spread]
|===
| *Type* a| <<QueryJobInfo>>
|===

=== For Configurations

* <<SfdcConfig>>

=== Throws

* SALESFORCE:CONNECTIVITY
* SALESFORCE:INVALID_REQUEST_DATA
* SALESFORCE:INVALID_SESSION
* SALESFORCE:INVALID_STRUCTURE_FOR_INPUT_DATA
* SALESFORCE:RETRY_EXHAUSTED
* SALESFORCE:UNKNOWN

[[BatchInfo]]
== Batch Info
`<salesforce:batch-info>`

> Access latest `BatchInfo` of a submitted `BatchInfo`. Allows tracking of the execution status.

Reviewer: Access the information for the latest submitted batch.  ??

| Content type a| Enumeration, one of:

** CSV
** JSON
** XML
** ZIP_CSV
** ZIP_JSON
** ZIP_XML |  Content type used at job creation. If not provided, the default value used is `ContentType.XML`. |  |
| Target Variable a| String |  The name of a variable that stores the operation's output |  |
| Target Value a| String |  An expression that evaluates the operation's output and stores the output in Target Variable |  #[payload] |
| Reconnection Strategy a| * <<reconnect>>
* <<reconnect-forever>> |  A retry strategy in case of connectivity errors |  |
|===

=== Output

[%autowidth.spread]
|===
| *Type* a| <<BatchInfoType>>
|===

=== For Configurations

* <<SfdcConfig>>

=== Throws

* SALESFORCE:CONNECTIVITY
* SALESFORCE:INVALID_REQUEST_DATA
* SALESFORCE:INVALID_SESSION
* SALESFORCE:INVALID_STRUCTURE_FOR_INPUT_DATA
* SALESFORCE:RETRY_EXHAUSTED
* SALESFORCE:UNKNOWN

[[BatchInfoList]]
== Batch Info List
`<salesforce:batch-info-list>`

Get information about all batches in a job.

=== Parameters

[%header,cols="20s,25a,30a,15a,10a"]
|===
| Name | Type | Description | Default Value | Required
| Configuration | String | The name of the configuration to use. | | x

> | Job Id a| String |  id of the job that you want to retrieve batch information for |  #[payload] |

Reviewer: ```suggestion
| Job Id a| String |  ID of the job for which to retrieve batch information |  #[payload] |
```

|===
| Name | Type | Description | Default Value | Required
| Configuration | String | The name of the configuration to use. | | x
| Batch To Retrieve a| <<BatchInfoType>> |  The com.sforce.async.BatchInfo being monitored |  #[payload] |
| Streaming Strategy a| * <<RepeatableInMemoryStream>>
* <<RepeatableFileStoreStream>>
* non-repeatable-stream |  Configure if repeatable streams should be used and their behavior |  |
| Headers a| Object |  |  |
| Target Variable a| String |  The name of a variable that stores the operation's output |  |
| Target Value a| String |  An expression that evaluates the operation's output and stores the output in Target Variable |  #[payload] |
| Reconnection Strategy a| * <<reconnect>>
* <<reconnect-forever>> |  A retry strategy in case of connectivity errors |  |
|===

=== Output

[%autowidth.spread]
|===
| *Type* a| Binary
|===

=== For Configurations

* <<SfdcConfig>>

=== Throws

* SALESFORCE:CONNECTIVITY
* SALESFORCE:INVALID_REQUEST_DATA
* SALESFORCE:INVALID_SESSION
* SALESFORCE:INVALID_STRUCTURE_FOR_INPUT_DATA
* SALESFORCE:RETRY_EXHAUSTED
* SALESFORCE:UNKNOWN

[[BulkInfo]]
== Bulk Info
`<salesforce:bulk-info>`

> Access latest `BatchInfo` of a submitted `BatchInfo`. Allows tracking of the execution status.

Reviewer: I think this description might be a C&P error.  Probably should be "Access the information for the latest bulk BatchInfo submission.  ??

| Headers a| Object |  |  |
| Target Variable a| String |  The name of a variable that stores the operation's output |  |
| Target Value a| String |  An expression that evaluates the operation's output and stores the output in Target Variable |  #[payload] |
| Reconnection Strategy a| * <<reconnect>>
* <<reconnect-forever>> |  A retry strategy in case of connectivity errors |  |
|===

=== Output

[%autowidth.spread]
|===
| *Type* a| <<BulkOperationResult>>
|===

=== For Configurations

* <<SfdcConfig>>

=== Throws

* SALESFORCE:CONNECTIVITY
* SALESFORCE:INVALID_REQUEST_DATA
* SALESFORCE:INVALID_SESSION
* SALESFORCE:INVALID_STRUCTURE_FOR_INPUT_DATA
* SALESFORCE:RETRY_EXHAUSTED
* SALESFORCE:UNKNOWN

[[ChangeOwnPassword]]
== Change Own Password
`<salesforce:change-own-password>`

Changes the password of the user linked to the connector's configuration.

=== Parameters

[%header,cols="20s,25a,30a,15a,10a"]
|===
| Name | Type | Description | Default Value | Required
| Configuration | String | The name of the configuration to use. | | x
| Old Password a| String |  The old password to be changed |  | x

> | New Password a| String |  The new password to be changed |  | x

Reviewer: ```suggestion
| New Password a| String |  The new password |  | x
```

| Reconnection Strategy a| * <<reconnect>>
* <<reconnect-forever>> |  A retry strategy in case of connectivity errors |  |
|===

=== For Configurations

* <<SfdcConfig>>

=== Throws

* SALESFORCE:CONNECTIVITY
* SALESFORCE:INVALID_REQUEST_DATA
* SALESFORCE:INVALID_SESSION
* SALESFORCE:INVALID_STRUCTURE_FOR_INPUT_DATA
* SALESFORCE:RETRY_EXHAUSTED
* SALESFORCE:UNKNOWN

[[CloseJob]]
== Close Job
`<salesforce:close-job>`

> Closes an open Job given its ID.

Reviewer: ```suggestion
Closes an open job according to the specified job ID.
```

[[ConvertLead]]
== Convert Lead
`<salesforce:convert-lead>`

Converts a Lead into an Account, Contact, or (optionally) an Opportunity.

=== Parameters

[%header,cols="20s,25a,30a,15a,10a"]
|===
| Name | Type | Description | Default Value | Required
| Configuration | String | The name of the configuration to use. | | x
| Lead Convert Request a| <<LeadConvertRequest>> |  Information needed for lead conversion |  #[payload] |
| Headers a| Object |  |  |
| Target Variable a| String |  The name of a variable that stores the operation's output |  |
| Target Value a| String |  An expression that evaluates the operation's output and stores the output in Target Variable |  #[payload] |
| Reconnection Strategy a| * <<reconnect>>
* <<reconnect-forever>> |  A retry strategy in case of connectivity errors |  |
|===

=== Output

[%autowidth.spread]
|===
| *Type* a| <<LeadConvertResult>>
|===

=== For Configurations

* <<SfdcConfig>>

=== Throws

* SALESFORCE:CONNECTIVITY
* SALESFORCE:INVALID_REQUEST_DATA
* SALESFORCE:INVALID_SESSION
* SALESFORCE:INVALID_STRUCTURE_FOR_INPUT_DATA
* SALESFORCE:RETRY_EXHAUSTED
* SALESFORCE:UNKNOWN

[[Create]]
== Create
`<salesforce:create>`

> Adds one or more new records to your organization's data. When you map your objects to the input of this message processor keep in mind that they need to match the expected type of the object at Salesforce. Take the CloseDate of an Opportunity as an example, if you set that field to a string of value "2019-12-13" it will be sent to Salesforce as a string and operation will be rejected on the basis that CloseDate is not of the expected type. The proper way to actually map it is to generate a Java Date object, you can do so using Groovy expression evaluator as `#[groovy:Date.parse("yyyy-MM-dd", "2019-12-13")]`.

Reviewer: ```suggestion
Adds one or more new records to your organization's data. When you map your objects to the input of this message processor keep in mind that they need to match the expected type of the object at Salesforce. 

For example, if you set the CloseDate field of an Opportunity to a string of value of "2019-12-13" it will be sent to Salesforce as a string and the operation will be rejected on the basis that CloseDate is not of the expected type. The proper way to map it is to generate a Java Date object, which you can do  by using a Groovy expression evaluator as `#[groovy:Date.parse("yyyy-MM-dd", "2019-12-13")]`.
```

=== Parameters

[%header,cols="20s,25a,30a,15a,10a"]
|===
| Name | Type | Description | Default Value | Required
| Configuration | String | The name of the configuration to use. | | x
| Job info a| <<JobInfoType>> |  The JobInfo in which the batch will be created. |  | x
| Query a| String |  The query to be executed. |  #[payload] |
| Target Variable a| String |  The name of a variable that stores the operation's output |  |
| Target Value a| String |  An expression that evaluates the operation's output and stores the output in Target Variable |  #[payload] |
| Reconnection Strategy a| * <<reconnect>>
* <<reconnect-forever>> |  A retry strategy in case of connectivity errors |  |
|===

=== Output

[%autowidth.spread]
|===
| *Type* a| <<BatchInfoType>>
|===

=== For Configurations

* <<SfdcConfig>>

=== Throws

* SALESFORCE:CONNECTIVITY
* SALESFORCE:INVALID_REQUEST_DATA
* SALESFORCE:INVALID_SESSION
* SALESFORCE:INVALID_STRUCTURE_FOR_INPUT_DATA
* SALESFORCE:RETRY_EXHAUSTED
* SALESFORCE:UNKNOWN

[[CreateBatchStream]]
== Create Batch Stream
`<salesforce:create-batch-stream>`

> Creates a Batch using the given stream within the specified Job. The stream can have a CSV, JSON, XML, ZIP_CSV, ZIP_JSON, or ZIP_XML format. This call uses the Bulk API. The operation will be done in asynchronous fashion.

Reviewer: ```suggestion
Creates a Batch using the given stream within the specified Job. The stream can have a CSV, JSON, XML, ZIP_CSV, ZIP_JSON, or ZIP_XML format. This call uses the Bulk API. The operation will be done asynchronously. 
```

|===
| Name | Type | Description | Default Value | Required
| Configuration | String | The name of the configuration to use. | | x
| Job info a| <<JobInfoType>> |  The JobInfo in which the batch will be created. This parameter defaults to payload content. |  | x
| Stream a| Binary |  A stream containing the data. The stream can have a CSV, JSON, XML, ZIP_CSV, ZIP_JSON, ZIP_XML format. |  #[payload] |
| Headers a| Object |  |  |
| Target Variable a| String |  The name of a variable that stores the operation's output |  |
| Target Value a| String |  An expression that evaluates the operation's output and stores the output in Target Variable |  #[payload] |
| Reconnection Strategy a| * <<reconnect>>
* <<reconnect-forever>> |  A retry strategy in case of connectivity errors |  |
|===

=== Output

[%autowidth.spread]
|===
| *Type* a| <<BatchInfoType>>
|===

=== For Configurations

* <<SfdcConfig>>

=== Throws

* SALESFORCE:CONNECTIVITY
* SALESFORCE:INVALID_REQUEST_DATA
* SALESFORCE:INVALID_SESSION
* SALESFORCE:INVALID_STRUCTURE_FOR_INPUT_DATA
* SALESFORCE:RETRY_EXHAUSTED
* SALESFORCE:UNKNOWN

[[CreateBulk]]
== Create Bulk
`<salesforce:create-bulk>`

> Adds one or more new records to your organization's data. This call uses the Bulk API. The creation will be done in asynchronous fashion.

Reviewer: ```suggestion
Adds one or more new records to your organization's data. This call uses the Bulk API. The creation operation is performed asynchronously.
```

=== Parameters

[%header,cols="20s,25a,30a,15a,10a"]
|===
| Name | Type | Description | Default Value | Required
| Configuration | String | The name of the configuration to use. | | x
| sObject Type a| String |  Type of object to create |  | x
| sObjects a| Array of Object |  An array of one or more sObject objects. |  #[payload] |
| Sobject Max Depth a| Number |  Async SObject recursive MAX_DEPTH check |  5 |
| Target Variable a| String |  The name of a variable that stores the operation's output |  |
| Target Value a| String |  An expression that evaluates the operation's output and stores the output in Target Variable |  #[payload] |
| Reconnection Strategy a| * <<reconnect>>
* <<reconnect-forever>> |  A retry strategy in case of connectivity errors |  |
|===

=== Output

[%autowidth.spread]
|===
| *Type* a| <<BatchInfoType>>
|===

=== For Configurations

* <<SfdcConfig>>

=== Throws

* SALESFORCE:CONNECTIVITY
* SALESFORCE:INVALID_REQUEST_DATA
* SALESFORCE:INVALID_SESSION
* SALESFORCE:INVALID_STRUCTURE_FOR_INPUT_DATA
* SALESFORCE:RETRY_EXHAUSTED
* SALESFORCE:UNKNOWN

[[CreateBulkForQuery]]
== Create Bulk For Query
`<salesforce:create-bulk-for-query>`

> Creates a Batch using the given query. This call uses the Bulk API. The operation will be done in asynchronous fashion.

Reviewer: ```suggestion
Creates a Batch using the given query. This call uses the Bulk API. This operation is performed asynchronously.
```

=== Parameters

[%header,cols="20s,25a,30a,15a,10a"]
|===
| Name | Type | Description | Default Value | Required
| Configuration | String | The name of the configuration to use. | | x
| Job info a| <<JobInfoType>> |  The JobInfo in which the batch will be created. |  | x
| Query a| String |  The query to be executed. |  #[payload] |
| Target Variable a| String |  The name of a variable that stores the operation's output |  |
| Target Value a| String |  An expression that evaluates the operation's output and stores the output in Target Variable |  #[payload] |
| Reconnection Strategy a| * <<reconnect>>
* <<reconnect-forever>> |  A retry strategy in case of connectivity errors |  |
|===

=== Output

[%autowidth.spread]
|===
| *Type* a| <<BulkOperationResult>>
|===

=== For Configurations

* <<SfdcConfig>>

=== Throws

* SALESFORCE:CONNECTIVITY
* SALESFORCE:INVALID_REQUEST_DATA
* SALESFORCE:INVALID_SESSION
* SALESFORCE:INVALID_STRUCTURE_FOR_INPUT_DATA
* SALESFORCE:RETRY_EXHAUSTED
* SALESFORCE:UNKNOWN

[[CreateJob]]
== Create Job
`<salesforce:create-job>`

> Creates a Job to perform one or more batches through Bulk API Operations.

Reviewer: ```suggestion
Creates a Job to perform one or more batches using Bulk API Operations.
```
